### PR TITLE
Massive memory usage optimization

### PR DESCRIPTION
**Pull request by bpblakely**
Optimized fast functions by only feeding the parallelized functions the necessary data required for the computation. Then merging the original dataframe afterwords (if required).

Note: Line 228 might be unnecessary, since unpack_json_fast already reduces the size of the dataframe. I don't have a clean way to measure the impact of memory usage during computation.

### CODE REVIEW

File: safegraph_py_functions/safegraph_py_functions.py [hunk 1]
```diff
@@ -218,20 +223,25 @@ def unpack_json_and_merge_fast(df, json_column='visitor_home_cbgs', key_col_name
...
+    chunks_list = [df_subset[i:i+chunk_n] for i in range(0,df_subset.shape[0],chunk_n)] 
     partial_explode_json = partial(explode_json_array, array_column=array_column, value_col_name= value_col_name, place_key= place_key,
                        file_key = file_key,array_sequence = array_sequence, zero_index = zero_index)
     with Pool() as pool:
         results = pool.map(partial_explode_json,chunks_list)
-    return pd.concat(results)
+    df_subset = pd.concat(results)
+    df_subset.drop([array_column,file_key],axis=1,inplace=True) # preparing to merge by dropping duplicates
+    
+    return df.merge(df_subset, on=[place_key])
```
Comment (ryanfoxsquire) [inline safegraph_py_functions/safegraph_py_functions.py]: @bpblakely  don't you need to merge by BOTH `place_key` and `file_key`? If you have passed in data from more than one file, then you the primary key for your data is [`place_key`, `file_key`]

File: safegraph_py_functions/safegraph_py_functions.py [hunk 2]
```diff
@@ -218,20 +223,25 @@ def unpack_json_and_merge_fast(df, json_column='visitor_home_cbgs', key_col_name
...
-    chunks_list = [df[i:i+chunk_n] for i in range(0,df.shape[0],chunk_n)] 
+    df_subset = df[[array_column,place_key,file_key]] # send only what we need
+    chunks_list = [df_subset[i:i+chunk_n] for i in range(0,df_subset.shape[0],chunk_n)] 
     partial_explode_json = partial(explode_json_array, array_column=array_column, value_col_name= value_col_name, place_key= place_key,
                        file_key = file_key,array_sequence = array_sequence, zero_index = zero_index)
     with Pool() as pool:
         results = pool.map(partial_explode_json,chunks_list)
-    return pd.concat(results)
+    df_subset = pd.concat(results)
+    df_subset.drop([array_column,file_key],axis=1,inplace=True) # preparing to merge by dropping duplicates
```
Comment (ryanfoxsquire) [inline safegraph_py_functions/safegraph_py_functions.py]: @bpblakely  I think you should only drop the `array_column` not the file_key (you need it to merge below)

File: safegraph_py_functions/safegraph_py_functions.py (hunk 1 above)
Comment (ryanfoxsquire) [reply safegraph_py_functions/safegraph_py_functions.py]: you could confirm by testing it on data from multiple files in the same dataframe.

**Review by ryanfoxsquire — COMMENTED** [review]
@bpblakely sorry for the delayed feedback. thanks for this PR,  this seems like a great improvement. 

Just a small question about how to join back at end of`explode_json_array_fast` (see inline comments)

File: safegraph_py_functions/safegraph_py_functions.py (hunk 1 above)
Comment (bpblakely) [reply safegraph_py_functions/safegraph_py_functions.py]: Thanks for catching that. I forgot to consider the case where the place_key could be a duplicate. After the change I confirmed it's working as intended/expected on both patterns and social distancing data.

**Comment by bpblakely** [comment]
@ryanfoxsquire Fixed the merge for explode_json_array_fast. Should be good to go.

**Pull request merged**